modules/Bio/EnsEMBL/Utils/Tree/Interval/Mutable/Node.pm

```diff
@@ -0,0 +1,664 @@
+  my $self = shift;
+
+  my ($left, $right) = ($self->left, $self->right);
+  
+  if (_height($left) - _height($right) >= 2) {
+    if (_height($left->left) >= _height($left->right)) {
+      # Left-Left case
+      $self->_right_rotate;
+      $self->_update_max_right_rotate;
+    } else {
+      # Left-Right case
+      $left->_left_rotate;
+      $self->_right_rotate;
+      $self->_update_max_right_rotate;
+    }
+  } elsif (_height($right) - _height($left) >= 2) {
+    if (_height($right->right) >= _height($right->left)) {
+      # Right-Right case
+      $self->_left_rotate;
+      $self->_update_max_left_rotate;
+    } else {
+      # Right-Left case
+      $right->_right_rotate;
+      $self->_left_rotate;
+      $self->_update_max_left_rotate;
+    }
+  }
+}
+
+sub _left_rotate {
+  my $self = shift;
+
+  my $right = $self->right;
+
+  $right->parent($self->parent);
+  unless (defined $right->parent) {
+    $self->tree->root($right);
+  } else {
+    if ($right->parent->left == $self) {
+      $right->parent->left($right);
+    } elsif ($right->parent->right == $self) {
+      $right->parent->right($right);
+    }
+  }
+  
+  $self->right($right->left);
+  $self->right->parent($self) if $self->right;
+  
+  $right->left($self);
+  $self->parent($right);
+
+  $self->_update_height;
+  $right->_update_height;
+  
+}
+
+sub _update_max_left_rotate { # handles Right-Right case and Right-Left case in rebalancing AVL tree
+  my $self = shift;
+
+  # update max of left sibling (x in first case, y in second)
+  my $parent = $self->parent;
+  my $parent_right = $parent->right;
+  if ($parent_right) {
+    my $parent_right_high = $parent_right->_highest_end;
+    if (!$parent_right->left and $parent_right->right) {
+      $parent_right->{max} = max $parent_right_high, $parent_right->right->{max};
+    } elsif ($parent_right->left and !$parent_right->right) {
+      $parent_right->{max} = max $parent_right_high, $parent_right->left->{max};
+    } elsif (!$parent_right->left and !$parent_right->right) {
+      $parent_right->{max} = $parent_right_high;
+    } else {
+      $parent_right->{max} = max $parent_right_high, $parent_right->left->{max}, $parent_right->right->{max};
+    }
+  }
+  
+  # update max of itself (z)
+  my $high = $self->_highest_end;
+  if (!$self->left and $self->right) {
+    $self->{max} = max $high, $self->right->{max};
+  } elsif ($self->left and !$self->right) {
+    $self->{max} = max $high, $self->left->{max};
+  } elsif (!$self->left and !$self->right) {
+    $self->{max} = $high;
+  } else {
+    $self->{max} = max $high, $self->left->{max}, $self->right->{max};
+  }
+  
+  # update max of parent (y in first case, x in second)
+  $parent->{max} = max $parent->left?$parent->left->{max}:0, $parent->right?$parent->right->{max}:0, $parent->_highest_end
+    if $parent;
+
+}
+
+sub _right_rotate {
+  my $self = shift;
+
+  my $left = $self->left;
+
+  $left->parent($self->parent);
+  unless (defined $left->parent) {
```

Comment:
Could this not be swapped so the `if (defined $right->parent)` then perform what is in the `else` block